File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function(nums, k) {
    const SIZE = nums.length;
    const frequency = new Map();
    const bucket = new Map();
    
    for (const num of nums) {
        const count = frequency.get(num) || 0;
        frequency.set(num, count + 1);
    }
    
    for (let i = 0; i < SIZE + 1; i++) {
        bucket.set(i, []);
    }
    
    for (const [num, count] of frequency.entries()) {
        bucket.get(count).push(num);
    }
    
    let index = 0;
    const answer = [];
    for (let i = SIZE; i >= 0; i--) {
        for (const num of bucket.get(i)) {
            answer[index++] = num;
            if (index === k) {
                return answer;
            }
        }
    }
};

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.kt
class Solution {
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        val answer = IntArray(k);
        val bucket: MutableMap<Int, MutableList<Int>> = mutableMapOf()
        val freq: MutableMap<Int, Int> = mutableMapOf()

        for (i in 0..nums.size) {
            bucket.putIfAbsent(i, mutableListOf())
        }

        for (num in nums) {
            freq[num] = freq.getOrDefault(num, 0) + 1
        }

        for (entries in freq.entries) {
            val num = entries.key
            val count = entries.value
            bucket[count]?.add(num)
        }

        var i = 0
        for (num in nums.size downTo 0) {
            for (element in bucket[num]!!) {
                answer[i++] = element
                if (i == k) {
                    return answer
                }
            }
        }

        return answer
    }
}